File: duniverse/lambdasoup.0.6.4/test/performance/performance.ml
let measure runs name f =
  let rec run = function
    | 0 -> ()
    | n -> f (); run (n - 1)
  in

  let start_time = Unix.gettimeofday () in

  run runs;

  let duration = (Unix.gettimeofday ()) -. start_time in
  let average = duration /. (float_of_int runs) *. 1000000. in

  Printf.printf "%s: %.0f us\n" name average

open Soup

let () =
  let html = read_file "../pages/google.html" in

  measure 100 "parse" (fun () -> parse html |> ignore);
  let soup = parse html in

  let traverse = fun () -> soup |> descendants |> count in
  assert (traverse () > 100);
  measure 1000 "traverse" (fun () -> traverse () |> ignore);

  let selector = "form[action*=search]" in
  assert (soup $ selector |> name = "form");
  measure 1000 "select" (fun () -> soup $ selector |> ignore);

  let selector = "*" in
  assert (soup $$ selector |> count > 10);
  measure 1000 "select_all" (fun () -> soup $$ selector |> count |> ignore)

File: duniverse/lambdasoup.0.7.1/test/performance/performance.py
import time

from bs4 import BeautifulSoup

def measure(runs, name, f):
    generator = xrange(runs)

    start_time = time.time()

    for i in generator:
        f()

    duration = time.time() - start_time
    average = duration / runs * 1000000

    print "%s: %.0f us" % (name, average)

def main():
    html = open("test/pages/google.html").read()

    def parse():
        return BeautifulSoup(html)

    measure(1000, "parse", parse)
    soup = parse()

    def traverse():
        count = 0
        for node in soup.descendants:
            count += 1
        return count

    assert traverse() > 100
    measure(1000, "traverse", traverse)

    def select():
        return soup.select("form[action*=search]")[0]

    assert select().name == "form"
    measure(1000, "select", select)

    def select_all():
        count = 0
        for node in soup.select("*"):
            count += 1
        return count

    assert select_all() > 10
    measure(1000, "select_all", select_all)

if __name__ == "__main__":
    main ()
